CorsixTH/Lua/dialogs/edit_room.lua

```diff
@@ -1086,72 +1114,96 @@ local function checkDoorWalls(x, y, wall, has_swingdoor)
     dx = 1
     dy = 0
   end
-
+  local invalid_tile = 0
   if th:getCell(x, y, wall_num) % 0x100 ~= 0 then
-    return false
+    invalid_tile = 4
   end
 
   -- If it is a swing door there are two more locations to check.
   if has_swingdoor then
-    if th:getCell(x - dx, y - dy, wall_num) % 0x100 ~= 0 or
-        th:getCell(x + dx, y + dy, wall_num) % 0x100 ~= 0 then
-      return false
+    if th:getCell(x - dx, y - dy, wall_num) % 0x100 ~= 0 then
+      invalid_tile = invalid_tile | 2
+    end
+    if th:getCell(x + dx, y + dy, wall_num) % 0x100 ~= 0 then
+      invalid_tile = invalid_tile | 8
     end
   end
-  return true
+  return invalid_tile
 end
 
 --! Check whether the given tile can function as a door entry/exit tile.
 --!param xpos (int) X position of the tile.
 --!param ypos (int) Y position of the tile.
 --!param player_id (int) Player id owning the hospital.
---!param flag_names (array) If set, array with two additional required properties.
+--!param wall wall facing direction
+--!param wall (string) wall facing direction
+--!param world wall facing direction
 --!return Whether the tile is considered to be valid.
-local function validDoorTile(xpos, ypos, player_id, flag_names)
+local function validDoorTile(xpos, ypos, player_id, wall, world)
   local th = TheApp.map.th
-
   local tile_flags = th:getCellFlags(xpos, ypos)
-  if not (tile_flags.buildable or tile_flags.passable or tile_flags.owner == player_id) then return false end
-  if not flag_names then return true end
-  return tile_flags[flag_names[1]] and tile_flags[flag_names[2]]
+  if not (tile_flags.buildable and tile_flags.owner == player_id) then
+    -- any object will cause it to be blocked (ignore litter)
+    if not(tile_flags.thob == 0 or tile_flags.thob == 62) then return false end
+    if tile_flags.passable then
+      -- check for any object with a passable tile built within the door footprint
+      -- drinks machine and reception desk in corridors for example
+      for o in pairs(world:findAllObjectsNear(xpos, ypos, 1)) do
+        if o.object_type.id ~= "bench" then
+          for _, xy in ipairs(o:getWalkableTiles()) do
+            if xy[1] == xpos and xy[2] == ypos then
+              return false
```

Comment:
This looks a lot like your other function, maybe refactor it into its own function, and use that at both places?